gui/tiletypes.lua

```diff
@@ -0,0 +1,1338 @@
+            key='CUSTOM_L',
+            label='Variant:',
+            options=variant_list,
+            initial_option=-1,
+            on_change=self.on_change_variant,
+        },
+        widgets.Divider { frame={t=18}, frame_style_l=false, frame_style_r=false, },
+    }
+
+    self:changeStone(-1)
+    self:setVisibility(self.visible)
+end
+
+function TileConfig:openStonePopup()
+    local dialog = guimat.MaterialDialog {
+        frame_title = "Stone Types",
+        use_inorganic = false,
+        use_creature = false,
+        use_plant = false,
+        on_select = function(_,value) self:changeStone(value) end,
+    }:show()
+
+    dialog.subviews.list:setChoices(stone_list, 1)
+    dialog.context_str= "Stone"
+end
+
+function TileConfig:setStoneEnabled(bool)
+    self.stone_enabled = bool
+    if not bool then
+        self.subviews.vein_type_label:setOption(self.subviews.vein_type_label.initial_option)
+        self:changeStone(nil)
+    end
+    self.subviews.vein_type_label:updateOptionLabel()
+end
+
+function TileConfig:changeStone(stone_index)
+    local stone_option = stone_dict[-1]
+    if stone_index then
+        stone_option = stone_dict[stone_index]
+        self.on_change_stone(stone_option.value)
+    end
+
+    local label = self.subviews.stone_label
+    local base_label = copyall(label.label)
+    base_label.key = label.key
+    base_label.key_sep = label.key_sep
+    base_label.on_activate = label.on_activate
+    label:setText({
+        base_label,
+        { gap=1, text=stone_option.label, pen=stone_option.pen }
+    })
+end
+
+function TileConfig:setVisibility(visibility)
+    self.frame = visibility and { h=19 } or { h=0 }
+    self.visible = visibility
+end
+
+--================================--
+--||       TiletypeWindow       ||--
+--================================--
+-- Interface for editing tiles
+
+TiletypeWindow = defclass(TiletypeWindow, widgets.Window)
+TiletypeWindow.ATTRS {
+    name = "tiletype_window",
+    frame_title="Tiletypes",
+    frame=UI_AREA,
+    frame_inset={b=1, t=1},
+    screen=DEFAULT_NIL,
+    options_popup=DEFAULT_NIL,
+    cur_mode="place",
+    mode_description = "",
+    cur_shape=-1,
+    cur_mat=-1,
+    cur_special=-1,
+    cur_variant=-1,
+    first_point=DEFAULT_NIL, ---@type df.coord
+    last_point=DEFAULT_NIL, ---@type df.coord
+}
+
+function TiletypeWindow:init()
+    local makeUIChars = function(center_char1, center_char2)
+        return {
+            {218, 196,          196,          191},
+            {179, center_char1, center_char2, 179},
+            {192, 196,          196,          217},
+        }
+    end
+    local makeUIPen = function(border_pen, center_pen)
+        --border_pen = type(border_pen) == "table" and border_pen or {fg=border_pen}
+        --border_pen.tile_color=true
+        center_pen = type(center_pen) == "table" and center_pen or {fg=center_pen}
+        center_pen.tile_color=true
+        return {
+            border_pen,
```

Comment:
Yup, I used `tile_color=true` to have the `tile_overrides` ascii characters still be colored like in the ascii view. Otherwise they would always be the standard black and white.
Of course ideally I would be using actual sprites instead of colored characters, but it's what I was able to put together.

Reply:
sorry, I meant having a combination single- and double- index array. I've tested just the color (no array) and a double-indexed array. I never tried non-table elements in the array.

Reply:
Ah, yeah, it didn't used to be possible until I added in the support for different combinations of tables and non-tables in DFHack/dfhack#4622 / DFHack/dfhack#4624, I specifically made sure that that combination was possible cause I knew it would be useful to me :p